glib-macros/src/properties.rs

```diff
@@ -0,0 +1,527 @@
+                    )?;
+                    Ok(PropDesc::new(
+                        ident.as_ref().unwrap().clone(),
+                        ty.clone(),
+                        ReceivedAttrs::new(attrs),
+                    ))
+                })
+        })
+        .collect::<syn::Result<_>>()
+}
+
+/// Converts a glib property name to a correct rust ident
+fn name_to_ident(name: &syn::LitStr) -> syn::Ident {
+    format_ident!("{}", name.value().replace('-', "_"))
+}
+
+/// Changes `Self` to another concrete type, to make it work in a different scope
+fn change_self_type<T: ToTokens>(source: &T, ident: &str) -> TokenStream2 {
+    TokenStream2::from_str(&source.to_token_stream().to_string().replace("Self", ident)).unwrap()
+}
+
+fn getter_prototype(ident: &syn::Ident, ty: &syn::Type) -> TokenStream2 {
+    quote!(fn #ident(&self) -> <#ty as glib::Property>::Value)
+}
+fn setter_prototype(ident: &syn::Ident, ty: &syn::Type) -> TokenStream2 {
+    let ident = format_ident!("set_{}", ident);
+    quote!(fn #ident(&self, value: <#ty as glib::Property>::Value))
+}
+fn expand_getset_properties_def(props: &[PropDesc]) -> TokenStream2 {
+    let defs = props
+        .iter()
+        .flat_map(|p| {
+            let ident = name_to_ident(&p.name);
+            let getter = p.get.is_some().then(|| getter_prototype(&ident, &p.ty));
+            let setter = p.set.is_some().then(|| setter_prototype(&ident, &p.ty));
+            [getter, setter]
+        })
+        .flatten();
+    quote!(#(#defs;)*)
+}
+
+fn expand_getset_properties_impl(imp_type_ident: &syn::Ident, props: &[PropDesc]) -> TokenStream2 {
+    let defs = props.iter().map(|p| {
+        let ident = name_to_ident(&p.name);
+        let field_ident = &p.field_ident;
+        let ty = &p.ty;
+
+        let getter = p.get.as_ref().map(|mfn| {
+            let body = match (p.member.as_ref(), mfn) {
+                (None, MaybeCustomFn::Default) => quote!(
+                     self.imp().#field_ident.get(|x| x.to_owned())
+                ),
+                (Some(member), MaybeCustomFn::Default) => quote!(
+                     self.imp().#field_ident.get(|x| x.#member.to_owned())
+                ),
+                (_, MaybeCustomFn::Custom(custom_fn)) => {
+                    let custom_fn = change_self_type(custom_fn, &imp_type_ident.to_string());
+                    quote!((#custom_fn)(&self.imp()))
+                }
+            };
+            let fn_prototype = getter_prototype(&ident, ty);
+            quote!(#fn_prototype { #body })
+        });
+        let setter = p.set.as_ref().map(|mfn| {
+            let body = match (p.member.as_ref(), mfn) {
+                (None, MaybeCustomFn::Default) => quote!(
+                     self.imp().#field_ident.set(move |x| *x = value)
+                ),
+                (Some(member), MaybeCustomFn::Default) => quote!(
+                     self.imp().#field_ident.set(move |x| x.#member = value)
```

Comment:
These implementations seem to me like they would be incorrect in most cases, `set_property` has to be used here otherwise it will skip the notification and validation.